cmd/flipt/cloud.go

```diff
@@ -0,0 +1,383 @@
+
+	cmd.PersistentFlags().StringVarP(&cloud.url, "url", "u", "https://flipt.cloud", "Flipt Cloud URL")
+
+	loginCmd := &cobra.Command{
+		Use:   "login [flags]",
+		Short: "Authenticate with Flipt Cloud",
+		RunE:  cloud.login,
+		Args:  cobra.NoArgs,
+	}
+
+	cmd.AddCommand(loginCmd)
+
+	logoutCmd := &cobra.Command{
+		Use:   "logout [flags]",
+		Short: "Logout from Flipt Cloud",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if err := os.Remove(filepath.Join(userConfigDir, "cloud.json")); err != nil && !errors.Is(err, os.ErrNotExist) {
+				return fmt.Errorf("removing cloud auth token: %w", err)
+			}
+
+			fmt.Println("Logged out from Flipt Cloud.")
+			return nil
+		},
+		Args: cobra.NoArgs,
+	}
+
+	cmd.AddCommand(logoutCmd)
+
+	serveCmd := &cobra.Command{
+		Use:   "serve [flags]",
+		Short: "Serve your local instance via Flipt Cloud",
+		RunE:  cloud.serve,
+		Args:  cobra.NoArgs,
+	}
+
+	cmd.AddCommand(serveCmd)
+
+	return cmd
+}
+
+func (c *cloudCommand) login(cmd *cobra.Command, args []string) error {
+	var (
+		done        = make(chan struct{})
+		ctx, cancel = context.WithCancel(cmd.Context())
+		_, cfg, err = buildConfig(ctx)
+	)
+	defer cancel()
+
+	if err != nil {
+		return err
+	}
+
+	if !cfg.Experimental.Cloud.Enabled {
+		return errors.New("cloud feature is not enabled")
+	}
+
+	_, err = url.Parse(c.url)
+	if err != nil {
+		return fmt.Errorf("parsing cloud URL: %w", err)
+	}
+
+	ok, err := util.PromptConfirm("Open browser to authenticate with Flipt Cloud?", false)
+	if err != nil {
+		return err
+	}
+
+	if !ok {
+		return nil
+	}
+
+	flow, err := cloud.InitFlow()
+	if err != nil {
+		return fmt.Errorf("initializing flow: %w", err)
+	}
+
+	var g errgroup.Group
+
+	cloudAuthFile := filepath.Join(userConfigDir, "cloud.json")
+
+	g.Go(func() error {
+		tok, err := flow.Wait(ctx)
+		if err != nil {
+			return fmt.Errorf("waiting for token: %w", err)
+		}
+
+		cloudAuth := cloudAuth{
+			Token: tok,
+		}
+
+		cloudAuthBytes, err := json.Marshal(cloudAuth)
+		if err != nil {
+			return fmt.Errorf("marshalling cloud auth token: %w", err)
+		}
+
+		if err := os.WriteFile(cloudAuthFile, cloudAuthBytes, 0600); err != nil {
+			return fmt.Errorf("writing cloud auth token: %w", err)
+		}
+
+		fmt.Println("\n✓ Authenticated with Flipt Cloud!\nYou can now run commands that require cloud authentication.")
+
+		return nil
+	})
+
+	g.Go(func() error {
+		if err := flow.StartServer(nil); err != nil && !errors.Is(err, net.ErrClosed) {
+			return fmt.Errorf("starting server: %w", err)
+		}
+		close(done)
+		return nil
+	})
+
+	g.Go(func() error {
+		select {
+		case <-done:
+			cancel()
+		case <-ctx.Done():
+			if err := flow.Close(); err != nil && !errors.Is(err, context.Canceled) {
+				return err
+			}
+		}
+		return nil
+	})
+
+	g.Go(func() error {
+		url, err := flow.BrowserURL(fmt.Sprintf("%s/login/device", c.url))
+		if err != nil {
+			return fmt.Errorf("creating browser URL: %w", err)
+		}
+		return util.OpenBrowser(url)
+	})
```

Comment:
Just making a note for future refactoring: lots of goroutines here, which I think have little extra value.
I think we need only one for the blocking server start and then the rest could probably be synchronous with this command.

1. go start server
2. open browser (looks non-blocking, is that right?)
3. flow.Wait which observes context cancelled
4. observe token and write out file